ERCS/erc-7841.md

```diff
@@ -0,0 +1,550 @@
+
+```solidity
+/// @title Mailbox contract implementation for synchronous communication
+contract Mailbox {
+    // ... Constructor + other simple functions like chain_id().
+
+    /// @notice nested map: blockNum -> metadataDigest -> payload
+    /// @dev Easy cleanup by `delete inbox[block.number -1]`
+    mapping(uint256 => mapping(bytes32 => bytes)) inbox;
+    // Mapping to detect key collisions: metadataDigest -> writtenFlag
+    mapping(bytes32 => bool) outboxNullifier;
+
+    // These hash values are computed incrementally.
+    /// @notice Nested map: blockNum -> srcChainId -> H(...H(m_2 | H(m_1))..)
+    /// @dev Easy cleanup by `delete inboxDigest[block.number -1]`
+    mapping(uint256 => mapping(uint32 => bytes32)) inboxDigest;
+    /// @notice Nested map: blockNum -> destChainId -> H(...H(m_2 | H(m_1))..)
+    /// @dev Easy cleanup by `delete outboxDigest[block.number -1]`
+    mapping(uint256 => mapping(uint32 => bytes32)) outboxDigest;
+
+    /// @dev Given the metadata (Message struct without payload field) of a 
+    ///     message, derive the digest used as the dictionary key for inbox/outbox.
+    function getMetadataDigest(
+        uint32 srcChainId,
+        uint32 destChainId,
+        address srcAddress,
+        address destAddress,
+        uint256 uid
+    ) pure public returns (bytes32) {
+        return
+            keccak(
+                abi.encodePacked(
+                    srcChainId,
+                    destChainId,
+                    srcAddress,
+                    destAddress,
+                    uid
+                )
+            );
+    }
+
+    /// @notice Conceptual "cleanup/reset" of mailbox after each block since 
+    ///     sync msgs are received immediately.
+    function _resetMailbox() private {
+        delete inbox[block.number - 1];
+        delete inboxDigest[block.number - 1];
+        delete outboxDigest[block.number - 1];
+    }
+
+    /// @notice Send a message to another chain
+    function send(
+        uint32 destChainId,
+        address destAddress,
+        uint256 uid,
+        bytes memory payload
+    ) public {
+        bytes32 key = getMetadataDigest(
+            this.chain_id(),
+            destChainId,
+            bytes32(srcAddress),
+            bytes32(msg.sender),
+            uid
+        );
+
+        // Prevent overwriting the same key
+        require(!outboxNullifier[key]);
+        outboxNullifier[key] = true;
+
+        // Update the outbox digest
+        // digest' = H(digest | metadata | payload)
+        outboxDigest[block.number][this.chain_id()] = keccak256(
+            abi.encodePacked(
+                outboxDigest[block.number][this.chain_id()],
+                key,
+                m.payload
+            )
+        );
+    }
+
+    /// @dev This function is called by the Coordinator. It can only be called 
```

Comment:
`Coordinator` is not defined in this doc.

Reply:
We did define the "coordination protocol" (basically the off-chain mechanism that relay the messages) in our terminology section. While it's true we didn't explicitly define its role, it's a bit out-of-scope to fully and accurately describe what a coordinator does. 

that said, I agree maybe it doesn't hurt to add a simple description like: 
"coordinators are the entity that monitors the sending events on multiple rollups, determines their message payload and relay them to the destination by populating them into the correct inbox"

Reply:
on a side note, @philippecamacho do you think we should remove this comment/restriction. Didn't we discuss internally that we don't have to restrict the permission to call this function, as long as it's called once (by anyone). It should be fine?

Reply:
Yes, because the _coordinator_  builds the block so it is enough to ensure this function can only be called once per block.